File: src/_218.cpp
/*
LeetCode 218 - The Skyline Problem

O(n log n) sweepline algorithm
The sweepline algorithm itself is pretty short and elegant. (10-line code)
However, the pre-sorting needs significant attentions when tie appears, i.e.,
when two endpoints are sharing the same x-coordinate. We need to obey exactly the rules below.

1. Endpoints of left boundaries come before endpoints of right boundaries.
2. Among left endpoints that sharing the same x-coordinate, higher endpoints appear before lower endpoints.
3. Among right endpoints that sharing the same x-coordinate, lower endpoints appear before higher endpoints.



Some "nasty" examples.

+---+----+
|   |    |
|   |    |
+---+----+


+--------+
|        |
+-----+  |
|     |  |
+--+  |  |
|  |  |  |
+--+--+--+


+--------+
|        |
+--------+
|        |
+--------+
|        |
+--------+
*/
class Solution {
public:
    vector<pair<int, int>> getSkyline(vector<vector<int>>& buildings) {
        vector<pair<int, int>> coord;
        for (int i=0; i < buildings.size(); i++) {
            coord.push_back(make_pair(i, 0));
            coord.push_back(make_pair(i, 1));
        }
        sort(coord.begin(), coord.end(), [&buildings](pair<int, int> a, pair<int, int> b) {
            if (buildings[a.first][a.second] == buildings[b.first][b.second])
                if (a.second != b.second) return a.second < b.second;
                else if (a.second == 0) return buildings[a.first][2] > buildings[b.first][2];
                else return buildings[a.first][2] < buildings[b.first][2];
            else return buildings[a.first][a.second] < buildings[b.first][b.second];
        });

        multiset<int> set;
        set.insert(0);
        vector<pair<int, int>> ans;
        for (auto pair : coord) {
            int oldMax = *set.rbegin();
            if (pair.second == 0) set.insert(buildings[pair.first][2]);
            else set.erase(set.find(buildings[pair.first][2]));
            if (oldMax != *set.rbegin()) ans.push_back(make_pair(buildings[pair.first][pair.second], *set.rbegin()));
        }
        return ans;
    }
};

File: src/_218.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.TreeMap;

/**
 * LeetCode 218 - The Skyline Problem
 *
 * Same idea as the C++ solution, but written in Java.
 *
 * Say u is an element of a
 * Then, u is an endpoint point, where
 * u[0] is the x-coordinate,
 * u[1] is the y-coordinate,
 * u[2] = 0 if u is a left endpoint, u[2] = 1 if it is a right endpoint.
 */
public class _218 {
    public List<int[]> getSkyline(int[][] buildings) {
        List<int[]> ans = new ArrayList<>();
        List<int[]> a = new ArrayList<>(buildings.length * 2);
        for (int[] building : buildings) {
            a.add(new int[]{building[0], building[2], 0});
            a.add(new int[]{building[1], building[2], 1});
        }
        Collections.sort(a, (u, v) -> {
            if (u[0] != v[0]) return Integer.compare(u[0], v[0]);
            else if (u[2] != v[2]) return Integer.compare(u[2], v[2]);
            else if (u[2] == 0) return -Integer.compare(u[1], v[1]);
            else return Integer.compare(u[1], v[1]);
        });

        TreeMap<Integer, Integer> map = new TreeMap<>();
        map.put(0, 1);
        for (int[] i : a) {
            int preMax = map.lastKey();
            if (i[2] == 0) map.put(i[1], map.getOrDefault(i[1], 0) + 1);
            else {
                map.put(i[1], map.get(i[1]) - 1);
                if (map.get(i[1]) == 0) map.remove(i[1]);
            }
            if (preMax != map.lastKey()) ans.add(new int[]{i[0], map.lastKey()});
        }
        return ans;
    }
}
